### DISCUSSION TITLE
PATCH와 PUT 메서드의 차이

### DISCUSSION THREAD

**Discussion by yeonkkk** [woowacourse-study/2023-tal-study#65] [기술 면접 질답 모음집]
PATCH와 PUT 메서드의 차이에 대해 설명해주세요.

**Comment by yoondgu** [comment]
PATCH와 PUT은 둘 다 자원을 변경하는 요청에 대한 메서드입니다.  
PATCH는 /members/1 일 때 'id가 1인 회원의 등급을 변경'하는 것과 같이 자원에 대한 부분적인 교체를 하지만,  
PUT은 '회원 정보 변경'으로 회원 정보 수정 폼에서 모든 정보를 업데이트 요청하는 것과 같이 자원 전체에 대한 교체를 할 때 사용합니다.  
둘의 큰 차이는 멱등성입니다.  
PUT은 자원 자체를 완전히 교체해버리기 때문에 멱등합니다. (자원의 일부가 바뀌는 게 아님)  
PATCH는 PUT과 달리 멱등성을 보장할 수 없습니다. 
이름을 변경하는 것과 같은 부분 변경이 아니라, 값을 증가시키는 등의 부분 변경도 허용되기 때문입니다.

**Comment by yeonkkk** [comment]
HTTP PUT 메서드는 요청 페이로드를 사용해 새로운 리소스를 생성하거나, 대상 리소스를 나타내는 데이터를 대체합니다.
HTTP PATCH 메소드는 리소스의 부분적인 수정을 할 때에 사용됩니다.
또 다른 차이로는 멱등성이 있고, PUT은 멱등성을 가지지만, PATCH는 멱등성을 가지지 않습니다.

**Reply by yeonkkk** [reply]
ref.
[자원을 수정하는 HTTP 메서드 - PUT vs PATCH](https://tecoble.techcourse.co.kr/post/2020-08-17-put-vs-patch/). 
[Patch 메서드가 멱등이 아닌 이유](https://www.inflearn.com/questions/110644/patch-%EB%A9%94%EC%84%9C%EB%93%9C%EA%B0%80-%EB%A9%B1%EB%93%B1%EC%9D%B4-%EC%95%84%EB%8B%8C-%EC%9D%B4%EC%9C%A0)
